# ChromBridGE

Genomic translocations can be an unintended result of CRISPR-Cas genome editing which can potentially impact interpretability of research and have serious clinical implications. Next-generation sequencing assays can read the sequence of DNA fragments containing translocations, but these reads may also be chimeric reads resulting from sequencing or library preparation artifacts, and existing software tools poorly distinguish between these cases. 

To address this problem, we have developed a novel algorithm to precisely detect translocation sites in genome-edited data, and have incorporated this algorithm into a tool called ChromBridGE which can be used to identify precise transition points in reads arising from translocations. The tool leverages properties specific to CRISPR-Cas genome editing to discriminate likely translocations from noise. We applied ChromBridGE to simulated and real data, and show that ChromBridGE correctly identifies transition points even when not provided this information beforehand. In addition, ChromBridGE identifies several examples of reads that may appear to be translocations, but using our method are identified as unlikely products of CRISPR-Cas9 editing. Thus, ChromBridGE is an important tool in identifying and characterizing translocations in genome-editing data.
